src/rust/engine/src/externs/options.rs

```diff
@@ -146,11 +161,121 @@ impl PyConfigSource {
 #[pyclass]
 struct PyOptionParser(OptionParser);
 
-type RankedVal<T> = (T, isize);
+// The pythonic value of a dict-typed option.
+type PyDictVal = HashMap<String, PyObject>;
+
+// The derivation of the option value, as a vec of (value, vec of source ranks) tuples.
+// A scalar value will always have a single source. A list/dict value may have elements
+// appended across multiple sources.
+// In ascending rank order, so the last value is the final value of the option.
+type OptionValueDerivation<T> = Vec<(T, Vec<isize>)>;
+
+// A tuple (final value, rank of final value, optional derivation of value).
+//
+// Note: The final value and its rank could be computed from the derivation (see above),
+// but the full derivation is not itself computed in normal usage.
+// We could get rid of this tuple type by representing the final value and its rank as
+// a singleton derivation (in the case where we don't otherwise need the full derivation).
+// But that would allocate two unnecessary Vecs for every option.
+type OptionValue<T> = (Option<T>, isize, Option<OptionValueDerivation<T>>);
+
+// Condense list value derivation across sources, so that it reflects merges vs. replacements
+// in a useful way. E.g., if we merge [a, b] and [c], and then replace it with [d, e],
+// the derivation will show:
+//   - [d, e] (from command-line flag)
+//   - [a, b, c] (from env var, from config)
+fn condense_list_value_derivation<T: PartialEq>(
+    derivation: Vec<(Source, Vec<ListEdit<T>>)>,
+) -> OptionValueDerivation<Vec<T>> {
+    let mut ret: OptionValueDerivation<Vec<T>> = vec![];
+    let mut cur_group = vec![];
+    let mut cur_ranks = vec![];
+
+    // In this case, for simplicity, we always use the "inefficient" O(M*N) remover,
```

Comment:
I think I would like to ban options of type "list of float" and get rid of this whole silliness. But that's for another day.